src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -299,15 +300,52 @@
           (app (assoc req :body-string payload))
           (http/error-response (str "Missing required parameter 'payload'")))))))
 
+(defn consume-and-close
+  "Consume all data from input stream and then close"
+  [^java.io.InputStream req-stream]
+  (with-open [s req-stream]
+    (loop [stream-bytes (byte-array 128)]
+      (when-not (= -1 (.read s stream-bytes))
+        (recur stream-bytes)))))
+
+(defn fail-when-payload-too-large
+  "Middlware that will return a 413 failure when the content-length of
+  a POST is too large (more than half the max heap). Acts as a noop
+  when that is not the case"
+  [app max-command-failure? max-command-size]
+  (fn [req]
+    (if (= :post (:request-method req))
+      (let [length-in-bytes (request/content-length req)]
+
+        (if length-in-bytes
+          (log/infof "Processing command with a content-length of %s bytes" length-in-bytes )
+          (log/warn (str "No content length found for POST. "
+                         "POST bodies that are too large could cause memory-related failures.")))
+
+        (if (and length-in-bytes
+                 max-command-failure?
+                 (> length-in-bytes max-command-size))
+          (do
+            (log/warnf "content-length of command is %sMB and is larger than the maximum allowed command size of %sMB"
+                       (utils/bytes->mb length-in-bytes)
+                       (utils/bytes->mb max-command-size))
+            (consume-and-close (:body req))
+            {:status 413
+             :headers {}
+             :body "POST body too large"})
```

Comment:
How about "Command reject due to size exceeding max-command-size" or something equally helpful. The agent already gets the 413 and presents an error "Request Entity Too Large" to the user, it just looks like we are repeating ourselves.